libblkid/src/superblocks/luks.c

```diff
@@ -110,39 +112,99 @@ static bool luks_valid(struct luks2_phdr *header, const char *magic, uint64_t of
 	return true;
 }
 
-static int probe_luks(blkid_probe pr, const struct blkid_idmag *mag __attribute__((__unused__)))
+static int luks_find_header(blkid_probe pr, struct luks2_phdr **header, uint64_t *offset)
 {
-	struct luks2_phdr *header;
+	struct luks2_phdr *h;
 	size_t i;
 
-	header = (struct luks2_phdr *) blkid_probe_get_buffer(pr, 0, sizeof(struct luks2_phdr));
-	if (!header)
+	h = (struct luks2_phdr *) blkid_probe_get_buffer(pr, 0, sizeof(struct luks2_phdr));
+	if (!h)
 		return errno ? -errno : BLKID_PROBE_NONE;
 
-	if (luks_valid(header, LUKS_MAGIC, 0)) {
+	if (luks_valid(h, LUKS_MAGIC, 0)) {
 		/* LUKS primary header was found. */
-		return luks_attributes(pr, header, 0);
+		if (header)
+			*header = h;
+		if (offset)
+			*offset = 0;
+		return BLKID_PROBE_OK;
 	}
 
 	/* No primary header, scan for known offsets of LUKS2 secondary header. */
 	for (i = 0; i < ARRAY_SIZE(secondary_offsets); i++) {
-		header = (struct luks2_phdr *) blkid_probe_get_buffer(pr,
-			  secondary_offsets[i], sizeof(struct luks2_phdr));
+		h = (struct luks2_phdr *) blkid_probe_get_buffer(pr,
+		     secondary_offsets[i], sizeof(struct luks2_phdr));
 
-		if (!header)
+		if (!h)
 			return errno ? -errno : BLKID_PROBE_NONE;
 
-		if (luks_valid(header, LUKS_MAGIC_2, secondary_offsets[i]))
-			return luks_attributes(pr, header, secondary_offsets[i]);
+		if (luks_valid(h, LUKS_MAGIC_2, secondary_offsets[i])) {
+			if (header)
+				*header = h;
+			if (offset)
+				*offset = secondary_offsets[i];
+			return BLKID_PROBE_OK;
+		}
 	}
 
 	return BLKID_PROBE_NONE;
 }
 
+static int probe_luks(blkid_probe pr, const struct blkid_idmag *mag __attribute__((__unused__)))
+{
+	struct luks2_phdr *header;
+	uint64_t offset;
+	int ret;
+
+	ret = luks_find_header(pr, &header, &offset);
+	if (ret != BLKID_PROBE_OK)
+		return ret;
+
+	return luks_attributes(pr, header, offset);
+}
+
+static int probe_luks_opal(blkid_probe pr, const struct blkid_idmag *mag __attribute__((__unused__)))
+{
+	struct luks2_phdr *header;
+	uint64_t offset;
+	int version;
+	int ret;
+
+	ret = luks_find_header(pr, &header, &offset);
+	if (ret != BLKID_PROBE_OK)
+		return ret;
+
+	/* To be sure there is no other superblock preceding LUKS2 header */
+	if (offset != 0)
+		return BLKID_PROBE_NONE;
+
+	version = be16_to_cpu(header->version);
+
+	if (version != 2)
+		return BLKID_PROBE_NONE;
+
+	if (memcmp((unsigned char *) header->subsystem,
```

Comment:
memcmp argument is void, cast is not needed. (and it must check for trailing \0 as mentioned above).

Reply:
Ah, copy-paste from luks_attributes.